### PR TITLE
Add fluid block and flowing fluid

### PR DESCRIPTION
**Pull request by shedaniel**
Allowing making common fluids and fluid blocks.

Tutorial write-up later...

### CODE REVIEW

File: common/src/main/java/dev/architectury/core/fluid/ArchitecturyFluidProperties.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+package dev.architectury.core.fluid;
+
+import dev.architectury.registry.registries.RegistrySupplier;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.block.LiquidBlock;
+import net.minecraft.world.level.material.Fluid;
+import org.jetbrains.annotations.Nullable;
+
+public class ArchitecturyFluidProperties {
```
Comment (shedaniel) [inline common/src/main/java/dev/architectury/core/fluid/ArchitecturyFluidProperties.java]: Not too sure about this class, might merge into the attributes interface

**Comment by shedaniel** [comment]
There are more work to do, this is wip

With how this implementation would convert between arch fluid stack <-> loader stacks/variants at lot, it would be beneficial to change how fluid stacks work in arch, I'll figure this out tomorrow

**Review by MaxNeedsSnacks — APPROVED** [review]

File: forge/build.gradle [hunk 2]
```diff
@@ -49,6 +49,14 @@ shadowJar {
 
     configurations = [project.configurations.shadowCommon]
     classifier "dev-shadow"
+
+    // Replace classes with forge's version
+    exclude "dev/architectury/core/block/ArchitecturyLiquidBlock.class"
+    exclude "dev/architectury/core/fluid/ArchitecturyFlowingFluid.class"
+    exclude "dev/architectury/core/fluid/ArchitecturyFlowingFluid\$Source.class"
+    exclude "dev/architectury/core/fluid/ArchitecturyFlowingFluid\$Flowing.class"
```
Comment (Juuxel) [inline forge/build.gradle]:
```suggestion
    exclude 'dev/architectury/core/fluid/ArchitecturyFlowingFluid$Source.class'
    exclude 'dev/architectury/core/fluid/ArchitecturyFlowingFluid$Flowing.class'
```
a single-quote string doesn't need to escape $ afaik

**Review by MaxNeedsSnacks — APPROVED** [review]

**Comment by MaxNeedsSnacks** [comment]
It's... beautiful. It's immaculate. It's the worst of both worlds! I wholeheartedly approve!

**Review by MaxNeedsSnacks — APPROVED** [review]
Can confirm this works without issues now, see https://github.com/KubeJS-Mods/KubeJS/tree/1.18/experimental-kube-fluids for an implementation of custom fluids using this pull request (needs a local build currently)

File: common/src/main/java/dev/architectury/core/item/ArchitecturyBucketItem.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    
+    private static <T> T checkPlatform(T obj) {
+        if (Platform.isForge()) {
+            throw new IllegalStateException("This class should've been replaced on Forge!");
+        }
+        
+        return obj;
+    }
+    
+    public final Fluid getContainingFluid() {
```
Comment (Juuxel) [inline common/src/main/java/dev/architectury/core/item/ArchitecturyBucketItem.java]: Contained fluid? I don't think water contains a water bucket

**Review by Juuxel — APPROVED** [review]

**Pull request merged**